README.md
# mipcc2023

Code for MIP computational competition 2023 (C++).

The execution file mipcomp.sh already has commands to compile the code if the code is not compiled.

The compiled code is in cppcode/build directory.

Use experiment.cpp file for any experimental evaluation. Modify the mipcomp.sh accordingly.

Ignore the python code. It is strictly for analysis and comparision. It is not used for the competition.


cppcode/changes.txt
after 22 dec full
Limit the number of solutions to 5
Add last 5 solutions and one common solution (0.9 similarity)
No more eventhandler
stop 2 seconds before timeout (time penalty worse than gap penalty)
(big effect expected) var history set pseudocost count to 4
--- 25 dec full ---
Provide no solution
--- 26 dec small10 ---
Revert the no solution part
Solution hiniting is now a parameter. Impact factor 0.8
--- 26 dec full ---
Solution hiniting Impact factor 0.7
maxcuts impact factor 0.7
--- 27 dec full ---
Don't update scores for the first instance.
Force attempt all values till count 5 in parameters
--- 28 dec full ---
New branching rule. 20% with 0.9 cost_update_factor
Fix penalty on unsuccessful hint. (remove heuristic time from penalty)
hint success stats
--- 29 dec full ---
Remove variable from hint if value is out of bounds.
New rule: Only update count of last level branching var when updating costs.
New rule: Two way pscost updates. 
New rule: No history reset. Threshold for reliability is 40.
Hinting: Update true score only if the hint was successful. Update the other score otherwise.
New branching rule. 20% with 0.9 cost_update_factor
--- 30 dec full ---
No custom branching
--- 31 dec full 1 ---
Always provide solution hints
--- 31 dec full 2 ---
Solution hint parameter again.
pscost history reset to 3
--- 1 jan full 1 ---
pscost history reset to 5
--- 1 jan full 2 ---
store global history
history reset count is a parameter.
--- 2 jan full 1 ---
history reset to 4 (non parameter)
--- 2 jan full 2 ---
max_cuts_root is a parameter
--- 3 jan full 1 ---
Print [END] timestamp after updating the parameter scores (and some print statements)
Provide at least one complete solution.
--- 4 jan full 1 ---
Level pscost as branching rule.
Revert complete solution change.
--- 4 jan full 2 ---
Revert levelpscost branching
No restarts
Heuristics and branching statistics at the end
--- 5 jan full ---
Turn off non performing heuristics after index 25
--- 6 jan full ---
Max restarts as parameter 0,1
fix the random seed
--- 7 jan full ---
Max restarts back to zero (causes wrong convergence in other params) REVISIT
--- 7 jan full 2 ---
parameters: If final scores are within 0.05, pick randomly
First solve always with default parameters
--- 8 jan full ---
parameters: If final scores are within 0.03, pick randomly
--- 9 jan full ---
New information from testfiles
if only obj change -> history reset to 1
--- 10 jan full ---
Levelpscost different implementation.
--- 11 jan full ---
Turn off levelpscost.
if only obj change -> history reset to 3
turn off non performing heuristics after index 15
print and record presolve and cuts stats
--- 14 jan full ---
turn off non performing heuristics after index 25
turn off non performing presolve after index 15
turn off non performing cuts after index 25
--- 15 jan full ---
Use mt19937 for random number generation in param tuning.
--- 15 jan full 2---
Recompile scip without levelpscost added code.
--- 16 jan full ---
switch to pscost after index 4
--- 16 jan full 2---
(if only lhs/rhs has changed) switch to pscost after index 4
Explore without randomness (explore count 8)
Tune cuts after index 4
--- 17 jan full ---
parameters: If final scores are within standard dev, pick randomly
if only obj change -> history reset to 1
--- 17 jan full 2---
parameters: If final scores are within standard dev/10, pick randomly
--- 19 jan full ---
No special treatment for history reset for obj only change
--- 19 jan private only full ---
Max restarts as parameter 0,1
--- 22 jan full ---
Use full strong branching for first instance
--- 23 jan full ---
Instead of full strong branching, increase efforts on reliability branching
Update restart 1 score only if actually restarted.
--- 24 jan full ---
revert to 23 jan full
max restart is again 0.
scip v 8.0.3
--- 25 jan full ---
back to scip v 8.0.2
--- 26 jan base full ---
Conditional compilation in mipcomp.sh
completesol heuristic nodesofs to 5000 (default 500)
completesol heuristic solutions to -1 (default 5)
Provide 9 prev solutions.
--- 30 jan full ---
Reduce efforts for completesol heuristic if obj only change.
Reuse best 10 cuts.
--- 31 jan full ---
Revert cut reuse change.
O3 compilation
--- 10 feb full ---
revert O3 compilation
--- 11 feb full ---
Code cleanup.
Max restart tuning after index 20.
--- 12 feb full ---
revert max restart tuning
Turn off heuristics with time per best soln / timeout > 0.2
--- 13 feb full ---
Heuristic turn off at instance 20
--- 18 feb full ---
back to Heuristic turn off at instance 25
Use timeout - 1 as limit
--- 24 feb private ---
--- 28 feb full server ---
--- 1 mar base full server ---
Remove base_dir from instance file path

cppcode/main.cpp
#include <iostream>
#include <scip/scip.h>
#include <scip/scipdefplugins.h>
#include <scip/struct_var.h>
#include <scip/struct_history.h>
#include "scip/history.h"
#include "scip/sol.h"
#include "scip/struct_stat.h"
#include "scip/struct_scip.h"
#include <string>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>
#include <iomanip>
#include <stdio.h>
#include <time.h>
#include <random>

#include "var_history.h"
#include "solutions.h"
#include "parameters.h"

#define DEBUG false

using namespace std;

// Get current date/time, format is YYYY-MM-DDTHH:mm:ss
const std::string CurrentDateTime()
{
    time_t now = time(0);
    struct tm tstruct;
    char buf[80];
    tstruct = *localtime(&now);
    // Visit http://en.cppreference.com/w/cpp/chrono/c/strftime
    // for more information about date/time format
    strftime(buf, sizeof(buf), "%Y-%m-%dT%X", &tstruct);

    return buf;
}

struct HeuristicStats
{
    string name;
    long long n_calls = 0ll;
    int n_solns = 0;
    int n_best_solns = 0;
    double time_spent = 0.0;

    void printstat()
    {
        std::cout << setw(25) << left << name
                  << "calls " << setw(10) << left << n_calls
                  << " solns " << setw(5) << left << n_solns
                  << " bestsolns " << setw(5) << left << n_best_solns
                  << " time " << time_spent
                  << endl;
    }
};

struct BranchingStats
{
    string name;
    long long n_calls = 0ll;
    int max_depth = 0;
    double time_spent = 0.0;

    void printstat()
    {
        std::cout << setw(25) << left << name
                  << "calls " << setw(10) << left << n_calls
                  << " maxdepth " << setw(10) << left << max_depth
                  << " time " << time_spent
                  << endl;
    }
};

struct PresolveStats
{
    string name;
    long long n_calls = 0ll;
    int total_changes = 0;
    double time_spent = 0.0;

    void printstat()
    {
        std::cout << setw(25) << left << name
                  << "calls " << setw(10) << left << n_calls
                  << " changes " << setw(10) << left << total_changes
                  << " time " << time_spent
                  << endl;
    }
};

struct SpearatorStats
{
    string name;
    long long n_calls = 0ll;
    long long cuts_found = 0ll;
    long long cuts_added = 0ll;
    double time_spent = 0.0;

    void printstat()
    {
        std::cout << setw(25) << left << name
                  << "calls " << setw(10) << left << n_calls
                  << " cuts_found " << setw(10) << left << cuts_found
                  << " cuts_added " << setw(10) << left << cuts_added
                  << " time " << time_spent
                  << endl;
    }
};

SCIP_RETCODE execmain(int argc, const char **argv)
{
    string meta_file_path = argv[1];
    std::cout << meta_file_path << endl;
    int pos = meta_file_path.find("datasets");
    string base_dir = meta_file_path.substr(0, pos);
    std::cout << base_dir << endl;
    pos = meta_file_path.find_last_of("/");
    string meta_file_name = meta_file_path.substr(pos + 1, string::npos);
    pos = meta_file_name.find(".");
    string meta_file_name_wo_ext = meta_file_name.substr(0, pos);

    // Create solutions directory
    string command = "mkdir -p solutions/" + meta_file_name_wo_ext;
    system(command.c_str());

    int timeout = 0;
    bool obj_only_change = true;
    bool rhs_only_change = true;
    bool obj_change = false;
    vector<string> instances;
    ifstream meta_file(meta_file_path);
    if (meta_file.is_open())
    {
        string timeout_str;
        string line;
        getline(meta_file, line);
        stringstream ss(line);
        ss >> timeout_str >> timeout;
        std::cout << timeout << endl;
        getline(meta_file, line); // OBJ
        string obj_str = "[OBJ] ";
        std::string::size_type i = line.find(obj_str);
        if (i != std::string::npos)
            line.erase(i, obj_str.length());
        if (line != "-")
        {
            obj_change = true;
            rhs_only_change = false;
        }

        getline(meta_file, line); // LO
        string lo_str = "[LO] ";
        i = line.find(lo_str);
        if (i != std::string::npos)
            line.erase(i, lo_str.length());
        if (line != "-")
        {
            std::cout << "LO changed " << endl;
            obj_only_change = false;
            rhs_only_change = false;
        }

        getline(meta_file, line); // UP
        string up_str = "[UP] ";
        i = line.find(up_str);
        if (i != std::string::npos)
            line.erase(i, up_str.length());
        if (line != "-")
        {
            std::cout << "UP changed " << endl;
            obj_only_change = false;
            rhs_only_change = false;
        }

        getline(meta_file, line); // LHS
        string lhs_str = "[LHS] ";
        i = line.find(lhs_str);
        if (i != std::string::npos)
            line.erase(i, lhs_str.length());
        if (line != "-")
        {
            std::cout << "LHS changed " << endl;
            obj_only_change = false;
        }

        getline(meta_file, line); // RHS
        string rhs_str = "[RHS] ";
        i = line.find(rhs_str);
        if (i != std::string::npos)
            line.erase(i, rhs_str.length());
        if (line != "-")
        {
            std::cout << "RHS changed " << endl;
            obj_only_change = false;
        }

        getline(meta_file, line); // MAT
        string mat_str = "[MAT] ";
        i = line.find(mat_str);
        if (i != std::string::npos)
            line.erase(i, mat_str.length());
        if (line != "-")
        {
            std::cout << "MAT changed " << endl;
            obj_only_change = false;
            rhs_only_change = false;
        }

        while (getline(meta_file, line))
        {
            instances.push_back(line);
        }
        // DO NOT PRINT INSTANCE NAMES. IT SPOILS THE EVAL.
        meta_file.close();
    }

    SCIP *scip = nullptr;
    SCIP_READER *reader;
    SCIP_CALL(SCIPcreate(&scip)); // Creating the SCIP environment
    // include default plugins.
    SCIP_CALL(SCIPincludeDefaultPlugins(scip));
    std::cout << "SCIP Version: " << SCIPversion() << endl;
    // Creating the SCIP Problem.
    SCIP_CALL(SCIPcreateProbBasic(scip, "Reoptimization"));

    // disable scip output to stdout
    SCIPmessagehdlrSetQuiet(SCIPgetMessagehdlr(scip), TRUE);

    // Parameters to tune. Use fixed seed for reproducibility.
    // Provide prev solution?
    Parameter<bool> provide_hint(0.3, "provide_hint", /*seed=*/56166895);
    provide_hint.SetExploreCount(8); // Hints are not always converted.
    provide_hint.SetSwitchFlag(1 << 0);
    provide_hint.AddValue(true);
    provide_hint.AddValue(false);
    int hint_success = 0;
    int hint_total = 0;

    // Tune after index 4.
    Parameter<int> max_cuts(0.3, "max_cuts", /*seed=*/984321);
    max_cuts.SetExploreCount(8);
    max_cuts.SetSwitchFlag(1 << 1);
    max_cuts.AddValue(100);
    max_cuts.AddValue(0);

    // Tune after index 4.
    Parameter<int> max_cuts_root(0.3, "max_cuts_root", /*seed=*/5198413);
    max_cuts_root.SetExploreCount(8);
    max_cuts_root.SetSwitchFlag(1 << 2);
    max_cuts_root.AddValue(2000);
    max_cuts_root.AddValue(0);

    SCIP_RESULT *result;
    result = new SCIP_RESULT[3];

    SolutionPool solution_pool;
    VarHistories var_histories;
    if (obj_only_change)
    {
        // All previous solutions are feasible. Spend less time exploring them.
        solution_pool.SetNumHintSolns(5); // Default is 9.
    }

    std::map<string, HeuristicStats> heuristic_stats;
    std::map<string, BranchingStats> branching_stats;
    std::map<string, PresolveStats> presolve_stats;
    std::map<string, SpearatorStats> sepa_stats;

    for (int index = 0; index < instances.size(); ++index)
    {
        string instance = instances[index];
        int pos = instance.find_last_of('/');
        string instance_name = instance.substr(pos + 1, string::npos);
        string filename = instance;

        std::cout << "[INSTANCE] " << instance_name << "\n"
                  << std::flush;
        // Read in *.MPS file
        SCIP_CALL(SCIPreadMps(scip, reader, filename.c_str(), result, NULL, NULL,
                              NULL, NULL, NULL, NULL));

        // Print the time
        // system("echo -n \"[START] \"; date -Iseconds");
        std::cout << "[START] " << CurrentDateTime() << "\n"
                  << std::flush;

        // Use little less than the time limit to ensure that we have
        // sufficient time left for stats collection and solution writing.
        // TODO: use timeout -1
        SCIP_CALL(SCIPsetRealParam(scip, "limits/time", timeout - 1));

        if (index == 0)
        {
            // Solve first instance with pure strong branching.
            SCIP_CALL(SCIPsetIntParam(scip, "branching/fullstrong/priority", 40000)); // default 0
        }
        if (index > 0)
        {
            // Back to normal.
            SCIP_CALL(SCIPsetIntParam(scip, "branching/fullstrong/priority", 0)); // default 0
            // Increase the efforts of completesol heuristic.
            if (!obj_only_change)
            {
                // Regular efforts works if only the objective has changed. Otherwise, increase the effort.
                SCIP_CALL(SCIPsetLongintParam(scip, "heuristics/completesol/nodesofs", 5000)); // default 500
                SCIP_CALL(SCIPsetIntParam(scip, "heuristics/completesol/solutions", -1));      // default 5
            }
            if (index > 4)
            {
                // Tune some parameters after a delay.
                SCIP_CALL(SCIPsetIntParam(scip, "separating/maxcuts", max_cuts.GetBestValue()));
                SCIP_CALL(SCIPsetIntParam(scip, "separating/maxcutsroot", max_cuts_root.GetBestValue()));
            }
            if (rhs_only_change && index > 4)
            {
                // Pseudocosts are well trained if objective and bounds are not changed.
                SCIP_CALL(SCIPsetIntParam(scip, "branching/pscost/priority", 40000)); // default 2000
            }
            // Disable restarts. We are already using enough information from previous instances.
            SCIP_CALL(SCIPsetIntParam(scip, "presolving/maxrestarts", 0));
        }
        if (index >= 25)
        {
            // Turn off non performing heuristics
            for (auto heuristic : heuristic_stats)
            {
                int num_best_solns = max(1, heuristic.second.n_best_solns);
                double time_per_best_soln = heuristic.second.time_spent / num_best_solns;
                if (heuristic.second.n_solns == 0)
                {
                    string heuristic_param = "heuristics/" + heuristic.first + "/freq";
                    SCIP_CALL(SCIPsetIntParam(scip, heuristic_param.c_str(), -1));
                }
                else if (time_per_best_soln / timeout > 0.2)
                {
                    // and some not so effective ones.
                    string heuristic_param = "heuristics/" + heuristic.first + "/freq";
                    SCIP_CALL(SCIPsetIntParam(scip, heuristic_param.c_str(), -1));
                }
            }

            // // Turn off non performing separators
            for (auto separator : sepa_stats)
            {
                if (separator.second.cuts_added == 0 && separator.second.time_spent > 1.0)
                {
                    string spea_param = "separating/" + separator.first + "/freq";
                    SCIP_CALL(SCIPsetIntParam(scip, spea_param.c_str(), -1));
                }
            }
        }

        if (index >= 15)
        {
            // Turn off non performing presolves
            for (auto presolve : presolve_stats)
            {
                if (presolve.second.total_changes == 0 && presolve.second.time_spent > 2.0)
                {
                    string presolve_param = "presolving/" + presolve.first + "/maxrounds";
                    SCIP_CALL(SCIPsetIntParam(scip, presolve_param.c_str(), 0));
                }
            }
        }

        // Store the original model variables.
        SCIP_VAR **vars;
        vars = SCIPgetOrigVars(scip);
        int num_vars = SCIPgetNOrigVars(scip);
        std::vector<SCIP_VAR *> scip_variables;
        for (int i = 0; i < num_vars; ++i)
        {
            SCIP_VAR *var = vars[i];
            scip_variables.push_back(var);
        }
        if (index > 0)
        {
            if (provide_hint.GetBestValue())
            {
                solution_pool.AddToModel(scip, scip_variables);
            }
            // solution_pool.SetCurrentScipVars(&scip_variables);
            var_histories.AddToModel(scip, scip_variables);
        }

        // Solve
        SCIP_CALL(SCIPsolve(scip));
        // Write solution
        if (SCIPgetNSols(scip) <= 0)
        {
            continue;
        }
        SCIP_SOL *sol;
        sol = SCIPgetBestSol(scip);

        // Dual bound.
        double dual_bound = SCIPgetDualbound(scip);
        std::cout << "[DUALBOUND] " << std::fixed << std::setprecision(9)
                  << dual_bound << "\n"
                  << std::flush;
        // Add solution to file and pool.
        ofstream solution_file;
        solution_file.open("solutions/" + meta_file_name_wo_ext + "/" + instance_name + ".sol");
        for (SCIP_VAR *scip_var : scip_variables)
        {
            const string name = SCIPvarGetName(scip_var);
            const double val = SCIPgetSolVal(scip, sol, scip_var);
            solution_file << name << " " << std::fixed << std::setprecision(9)
                          << val << "\n"
                          << std::flush;
        }
        solution_file.close();
        Solution solution;
        solution.Populate(scip, scip_variables, sol);
        solution_pool.AddSolution(solution);
        var_histories.Populate(scip, scip_variables);

        // Record solve statistics
        // Compute as per competition rules.
        double primal_bound = SCIPgetPrimalbound(scip);
        // double relative_gap = SCIPgetGap(scip);
        // relative_gap = min(relative_gap, 1.0);
        double relative_gap = 1.0;
        if (SCIPisFinite(primal_bound) && SCIPisFinite(dual_bound) && primal_bound * dual_bound >= 0)
        {
            if (primal_bound == 0 && dual_bound == 0)
            {
                relative_gap = 0.0;
            }
            else
            {
                relative_gap = abs(primal_bound - dual_bound) / max(abs(primal_bound), abs(dual_bound));
            }
        }
        double time_score = SCIPgetSolvingTime(scip) / timeout;
        if (relative_gap > 1e-6)
        {
            time_score = 1.0;
        }
        double total_score = time_score + relative_gap;
        if (DEBUG)
        {
            double adjusted_score = total_score * (1 + 0.1 * index);
            double first_sol_gap = scip->stat->firstsolgap;
            double first_solution_val = scip->stat->firstprimalbound;
            double best_solution_val = SCIPgetPrimalbound(scip);
            double first_solution_dist = abs(first_solution_val - best_solution_val);
            double first_sol_time = scip->stat->firstprimaltime;
            std::cout << "Relative gap: " << relative_gap << endl;
            std::cout << "time_score: " << time_score << endl;
            std::cout << "first_solution_dist: " << first_solution_dist << endl;
            std::cout << "first_sol_time: " << first_sol_time << endl;
            std::cout << "total_score: " << total_score << endl;
            std::cout << "adjusted_score: " << adjusted_score << endl;
            std::cout << "Number of runs: " << scip->stat->nruns << endl;
        }

        // Record branching stats
        SCIP_BRANCHRULE **branch_rules = SCIPgetBranchrules(scip);
        int n_branch_rules = SCIPgetNBranchrules(scip);
        for (int i = 0; i < n_branch_rules; ++i)
        {
            int n_calls = SCIPbranchruleGetNLPCalls(branch_rules[i]);
            if (n_calls > 0)
            {
                string name = SCIPbranchruleGetName(branch_rules[i]);
                double time_spent = SCIPbranchruleGetTime(branch_rules[i]);
                branching_stats[name].name = name;
                branching_stats[name].n_calls += n_calls;
                branching_stats[name].time_spent += time_spent;
            }
        }

        // Record heuristic stats
        SCIP_HEUR **heuristics = SCIPgetHeurs(scip);
        int n_heuristics = SCIPgetNHeurs(scip);
        for (int i = 0; i < n_heuristics; ++i)
        {
            int n_calls = SCIPheurGetNCalls(heuristics[i]);
            if (n_calls > 0)
            {
                string name = SCIPheurGetName(heuristics[i]);
                heuristic_stats[name].name = name;
                heuristic_stats[name].n_calls += n_calls;
                heuristic_stats[name].n_solns += SCIPheurGetNSolsFound(heuristics[i]);
                heuristic_stats[name].n_best_solns += SCIPheurGetNBestSolsFound(heuristics[i]);
                heuristic_stats[name].time_spent += SCIPheurGetTime(heuristics[i]);
            }
        }

        // Record Presolve stats
        SCIP_PRESOL **presolvers = SCIPgetPresols(scip);
        int n_presolvers = SCIPgetNPresols(scip);
        for (int i = 0; i < n_presolvers; ++i)
        {
            int n_calls = SCIPpresolGetNCalls(presolvers[i]);
            if (n_calls > 0)
            {
                string name = SCIPpresolGetName(presolvers[i]);
                presolve_stats[name].name = name;
                presolve_stats[name].time_spent += SCIPpresolGetTime(presolvers[i]);
                presolve_stats[name].n_calls += n_calls;
                int total_changes = SCIPpresolGetNChgBds(presolvers[i]) +
                                    SCIPpresolGetNAddConss(presolvers[i]) +
                                    SCIPpresolGetNAddHoles(presolvers[i]) +
                                    SCIPpresolGetNAggrVars(presolvers[i]) +
                                    SCIPpresolGetNChgCoefs(presolvers[i]) +
                                    SCIPpresolGetNChgSides(presolvers[i]) +
                                    SCIPpresolGetNChgVarTypes(presolvers[i]) +
                                    SCIPpresolGetNDelConss(presolvers[i]) +
                                    SCIPpresolGetNFixedVars(presolvers[i]) +
                                    SCIPpresolGetNUpgdConss(presolvers[i]);
                presolve_stats[name].total_changes += total_changes;
            }
        }

        // Record Cuts stats
        SCIP_SEPA **separators = SCIPgetSepas(scip);
        int n_sepas = SCIPgetNSepas(scip);
        for (int i = 0; i < n_sepas; ++i)
        {
            int n_calls = SCIPsepaGetNCalls(separators[i]);
            if (n_calls > 0)
            {
                string name = SCIPsepaGetName(separators[i]);
                sepa_stats[name].name = name;
                sepa_stats[name].n_calls += n_calls;
                sepa_stats[name].time_spent += SCIPsepaGetTime(separators[i]);
                sepa_stats[name].cuts_found += SCIPsepaGetNCutsFound(separators[i]);
                sepa_stats[name].cuts_added += SCIPsepaGetNCutsApplied(separators[i]);
            }
        }

        // Solution feedback
        SCIP_HEUR *comp_sol_heur = SCIPfindHeur(scip, "completesol");
        assert(comp_sol_heur != NULL);
        SCIP_Longint comp_sol_calls = SCIPheurGetNCalls(comp_sol_heur);
        SCIP_Longint comp_sol_solns = SCIPheurGetNSolsFound(comp_sol_heur);
        SCIP_Real comp_soln_time = SCIPheurGetTime(comp_sol_heur);
        if (DEBUG)
        {
            std::cout << "Calls " << comp_sol_calls
                      << " Solns " << comp_sol_solns
                      << " Time " << comp_soln_time
                      << endl;
        }
        if (comp_sol_calls > 0)
        {
            hint_total++;
            if (comp_sol_solns > 0)
            {
                hint_success++;
            }
        }

        // Update parameter scores.
        if (index > 0)
        {
            if (provide_hint.GetCurrentIndex() == 0) // True value for hint.
            {
                if (comp_sol_solns > 0)
                {
                    provide_hint.AdjustScore(-total_score);
                }
                else
                {
                    // Update the false value of hint if hint is not successful.
                    provide_hint.AdjustScore(-total_score, 1);
                }
            }
            else
            {
                provide_hint.AdjustScore(-total_score);
            }
            // Cut tuning starts at index 4, but we can still
            // update the scores as the first value is the default value.
            max_cuts.AdjustScore(-total_score);
            max_cuts_root.AdjustScore(-total_score);
        }
        // system("echo -n \"[END] \";date -Iseconds");
        std::cout << "[END] " << CurrentDateTime() << "\n\n"
                  << std::flush;
    }
    provide_hint.PrintStats();
    max_cuts.PrintStats();
    max_cuts_root.PrintStats();
    std::cout << "Provided hints: " << hint_total
              << " successful hints: " << hint_success << endl;

    for (auto branching : branching_stats)
    {
        branching.second.printstat();
    }

    for (auto heuristic : heuristic_stats)
    {
        heuristic.second.printstat();
    }

    for (auto presolve : presolve_stats)
    {
        presolve.second.printstat();
    }

    for (auto separator : sepa_stats)
    {
        separator.second.printstat();
    }

    return SCIP_OKAY;
}

int main(int argc, const char *argv[])
{
    return execmain(argc, argv) != SCIP_OKAY ? 1 : 0;
}
